hbase-common/src/main/java/org/apache/hadoop/hbase/util/FutureUtils.java

```diff
@@ -122,6 +120,41 @@ public static Throwable unwrapCompletionException(Throwable error) {
     return error;
   }
 
+  // This method is used to record the stack trace that calling the FutureUtils.get method. As in
+  // async client, the retry will be done in the retry timer thread, so the exception we get from
+  // the CompletableFuture will have a stack trace starting from the root of the retry timer. If we
+  // just throw this exception out when calling future.get(by unwrapping the ExecutionException),
+  // the upper layer even can not know where is the exception thrown...
+  // See HBASE-22316.
+  private static void setStackTrace(Throwable error) {
+    StackTraceElement[] localStackTrace = Thread.currentThread().getStackTrace();
+    StackTraceElement[] originalStackTrace = error.getStackTrace();
+    StackTraceElement[] newStackTrace =
+      new StackTraceElement[localStackTrace.length + originalStackTrace.length + 1];
+    System.arraycopy(localStackTrace, 0, newStackTrace, 0, localStackTrace.length);
+    newStackTrace[localStackTrace.length] =
+      new StackTraceElement("--------Future", "get--------", null, -1);
+    System.arraycopy(originalStackTrace, 0, newStackTrace, localStackTrace.length + 1,
+      originalStackTrace.length);
+    error.setStackTrace(newStackTrace);
+  }
+
+  private static IOException rethrow(ExecutionException error) throws IOException {
```

Comment:
The return parameter can be void?

Reply:
This is for the caller, we can just use `throw rethrow(e)`, otherwise you need to manually write an unused `throw new IOException()` at the end.